### ISSUE TITLE
Installation with dependencies should first download, then install

### ISSUE THREAD

**Issue by krlmlr** [r-lib/remotes#34]
I think installation of packages + dependencies should be split in three stages:
1. Download all source packages, including their dependencies
2. Install them in the correct order
3. Apply devtools-specific post-installation steps

For 2., we can leverage `install.packages()` with a `repos` argument that points to a local path: All we need to do is to download the packages to some `contrib/src/` subdirectory and call `tools::write_PACKAGES()`. This would finally give us parallel installation with `options(Ncpus > 1)`.

It looks like this would be relatively easy to implement with the current codebase of this repository.

Motivation: `remotes` does _a lot_ when installing packages and their dependencies, especially if these packages are from different remotes (e.g., CRAN + Bioconductor + GitHub). When revdep-checking RSQLite (with `bioconductor = TRUE`), I observed twice that I need 2-3 passes to install all revdeps. (The process is single-threaded and takes a day, I'm not too keen to repeat it _again_.)

Originally filed at https://github.com/hadley/devtools/issues/1308.

**Comment by gaborcsardi** [comment]
One complication with this approach is the `Remotes` field. You would need to extract the package, to check `DESCRIPTION` in case there is a `Remotes` field there. And then `Remotes` fields can trigger new downloads again.

**Comment by krlmlr** [comment]
Yes, that's right. But you download each package only once anyway, and the downloads can be cached easily, even persisting across sessions (=local CRAN-like repo).

**Comment by gaborcsardi** [comment]
Sure. It is entirely possible, just not *that* trivial. `

**Comment by gaborcsardi** [comment]
Some thoughts: 

* It would be great to have parallel downloads as well, and that is not entirely trivial. We would need to use `parallel` manually, and even the API that is not exported.
* It would be nice to have a command that does not download anything, but it gives you the dependency tree.
* Can we make the `Remotes` field easier to handle? E.g. we can put it in a DB with the commit hash included, and then if it is in the DB, we don't need to download the package to look at its `Remotes` field.

Ideally, we would do the downloads in parallel, and for non-CRAN-like packages we would extract `DESCRIPTION` as soon as the package is downloaded, to see if `Remotes` is there.

**Comment by krlmlr** [comment]
Can futures help? I wouldn't want to overcomplicate things with parallel downloads, the last mile is often the bottleneck anyway.

Which package wins in a diamond dependency structure ABD ACD if only B but not C asks to install D from a specific remote?

**Comment by gaborcsardi** [comment]
> Can futures help? I wouldn't want to overcomplicate things with parallel downloads, the last mile is often the bottleneck anyway.

Yes, but `remotes` should have no dependencies. That's why I mentioned `parallel`.

> Which package wins in a diamond dependency structure ABD ACD if only B but not C asks to install D from a specific remote?

IMO the `Remotes` one wins. However, if different remotes are requested, that should be an error imo. But these cases are rare, I think, `Remotes` is a temporary workaround before submitting to some CRAN-like repo, as I see it.

**Comment by krlmlr** [comment]
I don't think we can avoid downloading the same package several times in the general case. But we should be able to avoid installing it more than once -- and that's what this issue is (also) about.

**Comment by gaborcsardi** [comment]
We can definitely avoid downloading twice, at least on GH. E.g. we can look at `DESCRIPTION` with the GH API, without downloading the file. Whether it is worth it is another question, for a small package, it might not be faster than the download.

Here is an idea to avoid duplicate installs and downloads. We look for the non-CRAN-like package(s), and sort those out first (either via looking at `DESCRIPTION` or downloading them), and we do this recursively. In most cases, at most the top-level package (the one requested to install directly) will have `Remotes`, so this can be sorted out quickly. 

Once we have these special cases, the rest is easy and we download and install packages in parallel.

**Comment by krlmlr** [comment]
I was considering the most general case where DESCRIPTION is opaque until the entire .tar.gz is downloaded.

I'd still prefer if the download step was separate. This would also resolve your dependency tree wish. Here's how I think this can work:

- Create a task queue and a (temporary) local CRAN-like repo
- Add the package(s) to be installed to the task queue
- For each item fetched from the task queue, download the package to the CRAN-like repo and add its dependencies to the task queue
- Record downloaded package versions along the way, watch out for conflicts

In this very generic setting, we can of course give top-level packages with "Remotes" a higher priority. If we have an oracle that gives us the dependencies (via GitHub API or available.packages() or by parsing the CRAN page or ...), we may be able to save a download from CRAN, but that's not top priority IMO.

**Comment by gaborcsardi** [comment]
I think this is nice for separation, and it is also very simple to implement. It is basically almost what `remotes` is doing currently AFAIR. 

But this method limits very much what you can do in parallel.

E.g. in theory, if we work out the dependencies up front, then we can start installing packages as soon as the first package is downloaded. Working out the dependencies is easy, except for `Remotes`. `available.packages()` or really just a call to `http://crandeps.r-pkg.org/deps/{pkg}`.

I would prefer a method that works out the dependency tree without downloading too much stuff. E.g. just getting `DESCRIPTION` files will help me with system requirements, as I need to work these out before the Docker container is set up.

**Comment by krlmlr** [comment]
We can start installing packages as soon as the first *leaf* is downloaded, right. (After we prove that this is indeed a leaf, and not changed by a "Remotes" declaration which may introduce a new dependency.) For the very general case this amounts to knowing the entire dependency tree up front, so, yes. Can we discuss this in a separate GH issue?

I need to look again at what "remotes" is currently doing.

**Comment by krlmlr** [comment]
I've noticed that downloading multiple CRAN packages over the same HTTP(S) connection is ~4x faster for common package sizes, this may be another way to speed up things.

**Comment by gaborcsardi** [comment]
Can you download files over the same connection with `download.file()`?

**Comment by krlmlr** [comment]
At least "sometimes": You can pass a vector as url with method = "libcurl". I'll get back with some timing data in a second.

**Comment by gaborcsardi** [comment]
Wow, I didn't know that!

**Comment by krlmlr** [comment]
```r
library(magrittr)
N <- 20
df <- available.packages(repos = "https://cloud.r-project.org") %>%
  tibble::as_tibble() %>%
  head(N)

file_names <- paste0(df$Package, "_", df$Version, ".tar.gz")
urls <- paste0(df$Repository, "/", file_names)
destfiles <- file.path(tempdir(), file_names)
system.time(download.file(urls, file_names, method = "libcurl"))
system.time(Map(urls, file_names, f = download.file))
```

5 seconds with libcurl and parallel, 10 seconds with serial.

**Comment by gaborcsardi** [comment]
Do you know if it is documented that the libcurl method can take a vector?

**Comment by krlmlr** [comment]
Certainly is, but I don't know as of which version of R.

**Comment by krlmlr** [comment]
N = 100: 23 s vs. 42 s. Timing with curl::download_file() now.

**Comment by gaborcsardi** [comment]
Btw. we just added async multi-downloads to curl with Jeroen. Will need to iron out some details, but it will be totally possible to start installing the first package as soon as it is downloaded, while the others are still downloading.

But that is only if we have curl available, and `remotes` should work without `curl` as well, with a simpler download scheme.

**Comment by krlmlr** [comment]
If downloading happened in a separate R process, is there a reliable way to detect that the file is ready?

**Comment by gaborcsardi** [comment]
The other process must signalling it, e.g. by creating another file.

**Comment by gaborcsardi** [comment]
But the async cult multi-download runs in the same process, that's the cool thing about it.

**Comment by krlmlr** [comment]
CC @hadley

**Comment by krlmlr** [comment]
CC @jimhester

**Comment by krlmlr** [comment]
I guess this would also enable us to fix `install_deps(type = "both")` on Windows and OS X (it's currently silently overridden to `type = "binary"` because otherwise questions would be asked in interactive mode).

@gaborcsardi: Could we use crancache to set up a temporary repository and store the downloaded packages there?

**Comment by gaborcsardi** [comment]
This will not happen in remotes. pkgman will solve this.
